spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeletesRewrite.java

```diff
@@ -0,0 +1,429 @@
+      this.positionDelete = PositionDelete.create();
+      this.io = table.io();
+      this.specs = table.specs();
+
+      Types.StructType partitionType = Partitioning.partitionType(table);
+
+      this.specIdOrdinal = dsSchema.fieldIndex(PositionDeletesTable.SPEC_ID);
+      this.partitionOrdinalOption =
+          dsSchema.getFieldIndex(PositionDeletesTable.PARTITION).map(a -> (Integer) a);
+      this.partitionRowWrapper = initPartitionRowWrapper(partitionType);
+      this.partitionProjection =
+          StructProjection.create(partitionType, table.specs().get(specId).partitionType());
+
+      this.fileOrdinal = dsSchema.fieldIndex(MetadataColumns.DELETE_FILE_PATH.name());
+      this.positionOrdinal = dsSchema.fieldIndex(MetadataColumns.DELETE_FILE_POS.name());
+
+      this.rowOrdinal = dsSchema.fieldIndex(MetadataColumns.DELETE_FILE_ROW_FIELD_NAME);
+      DataType type = dsSchema.apply(MetadataColumns.DELETE_FILE_ROW_FIELD_NAME).dataType();
+      Preconditions.checkArgument(
+          type instanceof StructType, "Expected row as struct type but was %s", type);
+      this.rowSize = ((StructType) type).size();
+    }
+
+    @Override
+    public void write(InternalRow record) throws IOException {
+      int specId = record.getInt(specIdOrdinal);
+      PartitionSpec spec = specs.get(specId);
+
+      InternalRow partition = null;
+      if (partitionOrdinalOption.isDefined()) {
+        int partitionOrdinal = partitionOrdinalOption.get();
+        partition = record.getStruct(partitionOrdinal, partitionRowWrapper.size());
+      }
+      partitionProjection.wrap(partitionRowWrapper.wrap(partition));
+
+      String file = record.getString(fileOrdinal);
+      long position = record.getLong(positionOrdinal);
+      InternalRow row = record.getStruct(rowOrdinal, rowSize);
+      if (row != null) {
+        positionDelete.set(file, position, row);
+        lazyWriterWithRow().write(positionDelete, spec, partitionProjection);
+      } else {
+        positionDelete.set(file, position, null);
+        lazyWriterWithoutRow().write(positionDelete, spec, partitionProjection);
+      }
+    }
+
+    @Override
+    public WriterCommitMessage commit() throws IOException {
+      close();
+
+      List<DeleteFile> allDeleteFiles = Lists.newArrayList();
+      if (writerWithRow != null) {
+        allDeleteFiles.addAll(writerWithRow.result().deleteFiles());
+      }
+      if (writerWithoutRow != null) {
+        allDeleteFiles.addAll(writerWithoutRow.result().deleteFiles());
+      }
+      return new DeleteTaskCommit(allDeleteFiles);
+    }
+
+    @Override
+    public void abort() throws IOException {
+      close();
+
+      DeleteWriteResult resultWithRow = writerWithRow.result();
+      DeleteWriteResult resultWithoutRow = writerWithoutRow.result();
+      SparkCleanupUtil.deleteTaskFiles(
+          io,
+          Lists.newArrayList(
+              Iterables.concat(resultWithRow.deleteFiles(), resultWithoutRow.deleteFiles())));
```

Comment:
Can we do what you did in `commit` with an explicit list and if/else blocks?

```
List<DeleteFile> allDeleteFiles = Lists.newArrayList();
...
SparkCleanupUtil.deleteTaskFiles(io, allDeleteFiles);
```

This formatting is harder to read.